Index: node.h
===================================================================
RCS file: /cvsroot/firebird/interbase/dsql/node.h,v
retrieving revision 1.8
diff -U3 -r1.8 node.h
--- node.h	2001/08/19 03:49:21	1.8
+++ node.h	2001/10/13 09:01:57
@@ -344,9 +344,13 @@
     nod_udf_param, /* there should be a way to signal a param by descriptor! */
 
     /* limit support */
-	 nod_limit,
-	 nod_redef_procedure /* allows silent creation/overwriting of a procedure. */
+    nod_limit,
 
+    nod_redef_procedure, /* allows silent creation/overwriting of a procedure. */
+    
+    /* execute varchar support */
+    nod_exec_sql
+
 } NOD_TYPE;
 
 
@@ -480,6 +484,9 @@
 #define e_limit_skip	0	/* nod_limit */
 #define e_limit_length	1
 #define e_limit_count	2
+
+#define e_exec_vc		0   /* nod_exec_sql */
+#define e_exec_vc_count	1
 
 #define e_par_parameter	0	/* nod_parameter */
 #define e_par_count	1

