Index: par.c
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/par.c,v
retrieving revision 1.6
diff -U3 -r1.6 par.c
--- par.c	2001/10/11 04:25:56	1.6
+++ par.c	2001/10/13 09:19:54
@@ -2431,6 +2431,7 @@
     case blr_agg_total_distinct2:
     case blr_agg_average_distinct2:
     case blr_post:
+    case blr_exec_sql:
 	*arg++ = parse (tdbb, csb, sub_type);
 	break;
 

