Index: builds/original/prefix.solaris
===================================================================
RCS file: /cvsroot/firebird/interbase/builds/original/prefix.solaris,v
retrieving revision 1.2
diff -c -r1.2 prefix.solaris
*** builds/original/prefix.solaris	2000/12/14 19:29:42	1.2
--- builds/original/prefix.solaris	2001/12/04 14:34:26
***************
*** 14,19 ****
--- 14,23 ----
  #
  # All Rights Reserved.
  # Contributor(s): ______________________________________.
+ #
+ # 31-Oct-2001  Neil McCalden  Add large file support
+ #
+ 
  # Start of file prefix.solaris:	$(VERSION)	$(PLATFORM)
  .SUFFIXES: .c .e
  .e.c:
***************
*** 43,48 ****
--- 47,55 ----
  DEV_SHRLIB_DIR=		-L source/jrd
  DEV_VERSION_FLAG=	-DDEV_BUILD
  DEV_CFLAGS=		-g -xstrconst
+ 
+ # Also set SOLARIS_64IO in common.h
+ LARGEFILE_CFLAGS=`/usr/bin/getconf LFS_CFLAGS`
  
  DEBUG_OBJECTS=		$($(VERSION)_DEBUG_OBJECTS)
  SHRLIB_DIR=		$($(VERSION)_SHRLIB_DIR)
Index: burp/burp.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/burp.c,v
retrieving revision 1.7
diff -c -r1.7 burp.c
*** burp/burp.c	2001/08/09 00:27:39	1.7
--- burp/burp.c	2001/12/04 14:34:28
***************
*** 25,36 ****
   */
  
  
  
  #include "../jrd/ib_stdio.h"
  #include <stdlib.h>
  #include <string.h>
  #include <signal.h>
- #include "../jrd/common.h"
  #include "../jrd/ibase.h"
  #include <stdarg.h>
  #include "../jrd/ibsetjmp.h"
--- 25,42 ----
   */
  
  
+ #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
+ 
  #include <stdlib.h>
  #include <string.h>
  #include <signal.h>
  #include "../jrd/ibase.h"
  #include <stdarg.h>
  #include "../jrd/ibsetjmp.h"
***************
*** 1765,1771 ****
--- 1771,1781 ----
  	else
  #ifndef mpexl
  #ifndef WIN_NT
+ #ifdef SOLARIS_64IO
+ 	    if ((fil->fil_fd = open64 (fil->fil_name, MODE_WRITE, MASK)) == -1)
+ #else
  	    if ((fil->fil_fd = open (fil->fil_name, MODE_WRITE, MASK)) == -1)
+ #endif  
  #else
  	    if ((fil->fil_fd = MVOL_open (fil->fil_name, MODE_WRITE,
  		 CREATE_ALWAYS)) == INVALID_HANDLE_VALUE)
***************
*** 1914,1921 ****
--- 1924,1935 ----
      if ((fil->fil_fd =  MVOL_open (fil->fil_name, MODE_READ, OPEN_EXISTING))
  	 == INVALID_HANDLE_VALUE)
  #else
+ #ifdef SOLARIS_64IO
+     if ((fil->fil_fd = open64 (fil->fil_name, MODE_READ)) == INVALID_HANDLE_VALUE)		
+ #else
      if ((fil->fil_fd = open (fil->fil_name, MODE_READ)) == INVALID_HANDLE_VALUE)
  #endif
+ #endif
  	{
  	BURP_error (65, fil->fil_name, 0, 0, 0, 0);  /* msg 65 can't open backup file %s */
  	return QUIT;
***************
*** 1954,1961 ****
--- 1968,1980 ----
  	    if ((fil->fil_fd =  MVOL_open (fil->fil_name, MODE_READ,
  		 OPEN_EXISTING)) == INVALID_HANDLE_VALUE)
  #else
+ #ifdef SOLARIS_64IO
+ 	    if ((fil->fil_fd = open64 (fil->fil_name, MODE_READ))
+ 		 == INVALID_HANDLE_VALUE)		
+ #else			
  	    if ((fil->fil_fd = open (fil->fil_name, MODE_READ))
  		 == INVALID_HANDLE_VALUE)
+ #endif
  #endif
  		{
  #ifdef SUPERSERVER
Index: burp/mvol.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/mvol.c,v
retrieving revision 1.4
diff -c -r1.4 mvol.c
*** burp/mvol.c	2001/04/27 16:30:00	1.4
--- burp/mvol.c	2001/12/04 14:34:29
***************
*** 21,27 ****
--- 21,35 ----
   * Contributor(s): ______________________________________.
   */
  
+ #include "../jrd/common.h"
+ 
  #include "../jrd/ib_stdio.h"
+ 
+ /* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <stdlib.h>
  #include <string.h>
  #include <errno.h>
***************
*** 1018,1024 ****
--- 1026,1036 ----
      if((new_desc = MVOL_open(new_file,mode,OPEN_ALWAYS))
         == INVALID_HANDLE_VALUE)
  #else
+ #ifdef SOLARIS_64IO
+     if ((new_desc = open64 (new_file, mode, OPEN_MASK)) < 0)		
+ #else		
      if ((new_desc = open (new_file, mode, OPEN_MASK)) < 0)
+ #endif
  #endif /* WIN_NT */
  	{
  	BURP_print (222, new_file, 0, 0, 0, 0);
Index: burp/spit.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/spit.c,v
retrieving revision 1.2
diff -c -r1.2 spit.c
*** burp/spit.c	2000/08/30 04:21:35	1.2
--- burp/spit.c	2001/12/04 14:34:30
***************
*** 25,36 ****
   * Contributor(s): ______________________________________.
  */
  
  #include "../jrd/ib_stdio.h"
  #include <stdlib.h>
  #include <string.h>
  #include <errno.h>
  #include <assert.h>
- #include "../jrd/common.h"
  #include "../jrd/jrd.h"
  #include <stdarg.h>
  
--- 25,42 ----
   * Contributor(s): ______________________________________.
  */
  
+ #include "../jrd/common.h"
+ 
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <stdlib.h>
  #include <string.h>
  #include <errno.h>
  #include <assert.h>
  #include "../jrd/jrd.h"
  #include <stdarg.h>
  
***************
*** 667,674 ****
  
  		file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  		file_name	= fl_ptr->b_fil_name;
! 
  		output_fl_desc = open (file_name, MODE_WRITE, MASK);
  		if (output_fl_desc == -1)
  			{
  			free (io_buffer);
--- 673,683 ----
  
  		file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  		file_name	= fl_ptr->b_fil_name;
! #ifdef SOLARIS_64IO
! 		output_fl_desc = open64 (file_name, MODE_WRITE, MASK);
! #else
  		output_fl_desc = open (file_name, MODE_WRITE, MASK);
+ #endif
  		if (output_fl_desc == -1)
  			{
  			free (io_buffer);
***************
*** 746,753 ****
  
  						file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  						file_name	= fl_ptr->b_fil_name;
! 
  						output_fl_desc = open (file_name, MODE_WRITE, MASK);
  						if (output_fl_desc == -1)
  							{
  							free (io_buffer);
--- 755,765 ----
  
  						file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  						file_name	= fl_ptr->b_fil_name;
! #ifdef SOLARIS_64IO
! 						output_fl_desc = open64 (file_name, MODE_WRITE, MASK);
! #else
  						output_fl_desc = open (file_name, MODE_WRITE, MASK);
+ #endif
  						if (output_fl_desc == -1)
  							{
  							free (io_buffer);
***************
*** 1073,1079 ****
--- 1085,1095 ----
  TEXT		*char_ptr1, *char_ptr2, num_arr[5], total_arr[5];
  HEADER_REC	hdr_rec;
  
+ #ifdef SOLARIS_64IO
+ input_fl_desc = open64 (file_name, MODE_READ);
+ #else
  input_fl_desc = open (file_name, MODE_READ);
+ #endif
  
  if (input_fl_desc == -1)
  	{
Index: jrd/common.h
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/common.h,v
retrieving revision 1.18
diff -c -r1.18 common.h
*** jrd/common.h	2001/10/31 04:50:39	1.18
--- jrd/common.h	2001/12/04 14:34:31
***************
*** 253,267 ****
   */
  #define NEED_IB_STDIO
  
  /* Define the following macro only on platforms which supply the library
     routine xdr_hyper.  For other platforms, omit the definition of this
     macro, and we will supply that function where we need it. */
  #define PLATFORM_SUPPLIES_XDR_HYPER
- 
- #ifdef SOLX86
- #define LSEEK_OFFSET_CAST (off_t)
- #endif
- 
  
  #define ATEXIT(c)       atexit (c)
  #define SETPGRP         setpgrp ()
--- 253,272 ----
   */
  #define NEED_IB_STDIO
  
+ /* Used for modules which can not be compiled with standard largefile
+    cflags (see prefix.solaris|solx86) due to problems with ib_stdio.
+    This is primarly used for utilities which do their own direct db file 
+    access and output to screen. 
+    Hopefully this can be cleaned up in FB2.
+ */   
+ #ifdef sparc
+ #define SOLARIS_64IO
+ #endif
+ 
  /* Define the following macro only on platforms which supply the library
     routine xdr_hyper.  For other platforms, omit the definition of this
     macro, and we will supply that function where we need it. */
  #define PLATFORM_SUPPLIES_XDR_HYPER
  
  #define ATEXIT(c)       atexit (c)
  #define SETPGRP         setpgrp ()
Index: jrd/unix.c
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/unix.c,v
retrieving revision 1.6
diff -c -r1.6 unix.c
*** jrd/unix.c	2001/10/29 19:45:30	1.6
--- jrd/unix.c	2001/12/04 14:34:32
***************
*** 126,132 ****
--- 126,136 ----
  extern int	errno;
  
  static void	close_marker_file (TEXT *);
+ #if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+ static FIL	seek_file (FIL, BDB, off_t *, STATUS *);
+ #else
  static FIL	seek_file (FIL, BDB, SLONG *, STATUS *);
+ #endif
  static FIL	setup_file (DBB, TEXT *, USHORT, int);
  static BOOLEAN	unix_error (TEXT *, FIL, STATUS, STATUS *);
  
***************
*** 326,331 ****
--- 330,339 ----
  flag = O_RDWR | O_CREAT | (overwrite ? O_TRUNC : O_EXCL) | O_BINARY;
  #endif
  
+ #ifdef SOLARIS_64IO
+ flag = flag | O_LARGEFILE;
+ #endif
+ 
  if ((desc = open (file_name, flag, MASK)) == -1)
      ERR_post (isc_io_error,
  	gds_arg_string, "open O_CREAT",
***************
*** 675,680 ****
--- 683,692 ----
  flag = O_RDWR | O_BINARY;
  #endif
  
+ #ifdef SOLARIS_64IO
+ flag = flag | O_LARGEFILE;
+ #endif
+ 
  for (i = 0; i < IO_RETRY; i++)
      {
  
***************
*** 732,739 ****
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size, offset;
  
  ISC_inhibit();
  
  if (file->fil_desc == -1)
--- 744,757 ----
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size; 
  
+ #if defined (SOLARIS_64IO) && defined (SUPERSERVER)
+ off_t offset;
+ #else
+ SLONG  offset;
+ #endif
+ 
  ISC_inhibit();
  
  if (file->fil_desc == -1)
***************
*** 833,840 ****
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size, offset;
  
  ISC_inhibit();
  
  if (file->fil_desc == -1)
--- 851,864 ----
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size;
  
+ #if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+ 	off_t  offset;
+ #else
+ 	SLONG offset;
+ #endif
+ 
  ISC_inhibit();
  
  if (file->fil_desc == -1)
***************
*** 945,951 ****
--- 969,979 ----
  static FIL seek_file (
      FIL		file,
      BDB		bdb,
+ #if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+     off_t	*offset,
+ #else
      SLONG	*offset,
+ #endif
      STATUS	*status_vector)
  {
  /**************************************
***************
*** 982,997 ****
  page -= file->fil_min_page - file->fil_fudge;
  
  #ifdef PREAD_PWRITE
! *offset = (SLONG) (page * dbb->dbb_page_size);
  #else
  lseek_offset = page;
  lseek_offset *= dbb->dbb_page_size;
  #ifndef UNIX_64_BIT_IO
  if (lseek_offset > MAX_SLONG)
  {
      return (FIL) unix_error ("lseek", file, isc_io_32bit_exceeded_err, status_vector);
  }
  #endif
  
  THD_MUTEX_LOCK (file->fil_mutex);
  
--- 1010,1027 ----
  page -= file->fil_min_page - file->fil_fudge;
  
  #ifdef PREAD_PWRITE
! *offset = (off_t) (page * dbb->dbb_page_size);
  #else
  lseek_offset = page;
  lseek_offset *= dbb->dbb_page_size;
  #ifndef UNIX_64_BIT_IO
+ #ifndef _LARGEFILE_SOURCE
  if (lseek_offset > MAX_SLONG)
  {
      return (FIL) unix_error ("lseek", file, isc_io_32bit_exceeded_err, status_vector);
  }
  #endif
+ #endif
  
  THD_MUTEX_LOCK (file->fil_mutex);
  
***************
*** 1146,1152 ****
      int fd,
      SCHAR *buf,
      SLONG nbytes,
!     SLONG offset)
  /**************************************
   *
   *	p r e a d
--- 1176,1187 ----
      int fd,
      SCHAR *buf,
      SLONG nbytes,
! #ifdef SOLARIS_64IO
! 	off_t
! #else
!     SLONG offset
! #endif
! )
  /**************************************
   *
   *	p r e a d
***************
*** 1183,1189 ****
      int fd,
      SCHAR *buf,
      SLONG nbytes,
!     SLONG offset)
  /**************************************
   *
   *	p w r i t e
--- 1218,1229 ----
      int fd,
      SCHAR *buf,
      SLONG nbytes,
! #ifdef SOLARIS_64IO
! 	off_t offset
! #else
!     SLONG offset
! #endif
! )
  /**************************************
   *
   *	p w r i t e
Index: utilities/dba.e
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/dba.e,v
retrieving revision 1.10
diff -c -r1.10 dba.e
*** utilities/dba.e	2001/10/29 19:45:30	1.10
--- utilities/dba.e	2001/12/04 14:34:34
***************
*** 27,32 ****
--- 27,37 ----
  #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <errno.h>
  #include <string.h>
  #include <fcntl.h>
***************
*** 2250,2256 ****
--- 2255,2265 ----
  fil->fil_fudge = 0;
  fil->fil_max_page = 0L;
  
+ #ifdef SOLARIS_64IO
+ if ((fil->fil_desc = open64 (file_name, O_RDONLY)) == -1)
+ #else
  if ((fil->fil_desc = open (file_name, O_RDONLY)) == -1)
+ #endif
      {
  #ifdef SUPERSERVER
      CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
***************
*** 2316,2322 ****
  
  page_number -= fil->fil_min_page - fil->fil_fudge;
  offset = ((UINT64)page_number) * ((UINT64)tddba->page_size);
! if (lseek (fil->fil_desc, offset, 0) == -1)
      {
  #ifdef SUPERSERVER
      CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
--- 2325,2335 ----
  
  page_number -= fil->fil_min_page - fil->fil_fudge;
  offset = ((UINT64)page_number) * ((UINT64)tddba->page_size);
! #ifdef SOLARIS_64IO
! if (lseek64 (fil->fil_desc, offset, SEEK_SET) == -1)
! #else
! if (lseek (fil->fil_desc, LSEEK_OFFSET_CAST offset, 0) == -1)
! #endif
      {
  #ifdef SUPERSERVER
      CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
Index: utilities/rebuild.c
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/rebuild.c,v
retrieving revision 1.2
diff -c -r1.2 rebuild.c
*** utilities/rebuild.c	2001/10/29 19:45:30	1.2
--- utilities/rebuild.c	2001/12/04 14:34:35
***************
*** 23,28 ****
--- 23,33 ----
  #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <errno.h>
  
  #include "../jrd/gds.h"
***************
*** 658,665 ****
   *	Open a database file.
   *
   **************************************/
! 
  if ((rbdb->rbdb_file.fil_file = open (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
      db_error (errno);
  }
  
--- 663,673 ----
   *	Open a database file.
   *
   **************************************/
! #ifdef SOLARIS_64IO
! if ((rbdb->rbdb_file.fil_file = open64 (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
! #else
  if ((rbdb->rbdb_file.fil_file = open (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
+ #endif
      db_error (errno);
  }
  
***************
*** 685,691 ****
--- 693,703 ----
  file = rbdb->rbdb_file.fil_file;
  
  offset = ((UINT64)page_number) * ((UINT64)rbdb->rbdb_page_size);
+ #ifdef SOLARIS_64IO
+ if (lseek64 (file, offset, SEEK_SET) == -1)
+ #else
  if (lseek (file, offset, 0) == -1)
+ #endif
      db_error (errno);
  
  for (p = (SCHAR*) rbdb->rbdb_buffer1, length = rbdb->rbdb_page_size; length > 0;)
***************
*** 726,732 ****
--- 738,748 ----
  fd = rbdb->rbdb_file.fil_file;
  
  offset = ((UINT64)page_number) * ((UINT64)page_size);
+ #ifdef SOLARIS_64IO
+ if (lseek64 (fd, offset, SEEK_SET) == -1)
+ #else
  if (lseek (fd, offset, 0) == -1)
+ #endif
      db_error (errno);
  if (write (fd, page, page_size) == -1)
      db_error (errno);
Index: utilities/sbc_print.c
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/sbc_print.c,v
retrieving revision 1.3
diff -c -r1.3 sbc_print.c
*** utilities/sbc_print.c	2001/10/29 19:45:30	1.3
--- utilities/sbc_print.c	2001/12/04 14:34:36
***************
*** 23,28 ****
--- 23,33 ----
  #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <string.h>
  #include <fcntl.h>
  #include <errno.h>
***************
*** 858,864 ****
--- 863,873 ----
   *
   **************************************/
  
+ #idef SOLARIS_64IO
+ if ((file = open64 (file_name, 2)) == -1)
+ #else
  if ((file = open (file_name, 2)) == -1)
+ #endif
      db_error (errno);
  }
  
***************
*** 880,886 ****
--- 889,899 ----
  UINT64 offset;
  
  offset = ((UINT64)page_number) * ((UINT64)page_size);
+ #ifdef SOLARIS_64IO
+ if (lseek64 (file, offset, SEEK_SET) == -1)
+ #else
  if (lseek (file, offset, 0) == -1)
+ #endif
      db_error (errno);
  
  for (p = (SCHAR*) global_buffer, length = page_size; length > 0;)
Index: builds/original/prefix.solaris
===================================================================
RCS file: /cvsroot/firebird/interbase/builds/original/prefix.solaris,v
retrieving revision 1.2
diff -c -r1.2 prefix.solaris
*** builds/original/prefix.solaris	2000/12/14 19:29:42	1.2
--- builds/original/prefix.solaris	2001/12/05 00:30:39
***************
*** 14,19 ****
--- 14,23 ----
  #
  # All Rights Reserved.
  # Contributor(s): ______________________________________.
+ #
+ # 31-Oct-2001  Neil McCalden  Add large file support
+ #
+ 
  # Start of file prefix.solaris:	$(VERSION)	$(PLATFORM)
  .SUFFIXES: .c .e
  .e.c:
***************
*** 43,48 ****
--- 47,55 ----
  DEV_SHRLIB_DIR=		-L source/jrd
  DEV_VERSION_FLAG=	-DDEV_BUILD
  DEV_CFLAGS=		-g -xstrconst
+ 
+ # Also set SOLARIS_64IO in common.h
+ LARGEFILE_CFLAGS=`/usr/bin/getconf LFS_CFLAGS`
  
  DEBUG_OBJECTS=		$($(VERSION)_DEBUG_OBJECTS)
  SHRLIB_DIR=		$($(VERSION)_SHRLIB_DIR)
Index: burp/burp.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/burp.c,v
retrieving revision 1.7
diff -c -r1.7 burp.c
*** burp/burp.c	2001/08/09 00:27:39	1.7
--- burp/burp.c	2001/12/05 00:30:41
***************
*** 25,36 ****
   */
  
  
  
  #include "../jrd/ib_stdio.h"
  #include <stdlib.h>
  #include <string.h>
  #include <signal.h>
- #include "../jrd/common.h"
  #include "../jrd/ibase.h"
  #include <stdarg.h>
  #include "../jrd/ibsetjmp.h"
--- 25,42 ----
   */
  
  
+ #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
+ 
  #include <stdlib.h>
  #include <string.h>
  #include <signal.h>
  #include "../jrd/ibase.h"
  #include <stdarg.h>
  #include "../jrd/ibsetjmp.h"
***************
*** 1765,1771 ****
--- 1771,1781 ----
  	else
  #ifndef mpexl
  #ifndef WIN_NT
+ #ifdef SOLARIS_64IO
+ 	    if ((fil->fil_fd = open64 (fil->fil_name, MODE_WRITE, MASK)) == -1)
+ #else
  	    if ((fil->fil_fd = open (fil->fil_name, MODE_WRITE, MASK)) == -1)
+ #endif  
  #else
  	    if ((fil->fil_fd = MVOL_open (fil->fil_name, MODE_WRITE,
  		 CREATE_ALWAYS)) == INVALID_HANDLE_VALUE)
***************
*** 1914,1921 ****
--- 1924,1935 ----
      if ((fil->fil_fd =  MVOL_open (fil->fil_name, MODE_READ, OPEN_EXISTING))
  	 == INVALID_HANDLE_VALUE)
  #else
+ #ifdef SOLARIS_64IO
+     if ((fil->fil_fd = open64 (fil->fil_name, MODE_READ)) == INVALID_HANDLE_VALUE)		
+ #else
      if ((fil->fil_fd = open (fil->fil_name, MODE_READ)) == INVALID_HANDLE_VALUE)
  #endif
+ #endif
  	{
  	BURP_error (65, fil->fil_name, 0, 0, 0, 0);  /* msg 65 can't open backup file %s */
  	return QUIT;
***************
*** 1954,1961 ****
--- 1968,1980 ----
  	    if ((fil->fil_fd =  MVOL_open (fil->fil_name, MODE_READ,
  		 OPEN_EXISTING)) == INVALID_HANDLE_VALUE)
  #else
+ #ifdef SOLARIS_64IO
+ 	    if ((fil->fil_fd = open64 (fil->fil_name, MODE_READ))
+ 		 == INVALID_HANDLE_VALUE)		
+ #else			
  	    if ((fil->fil_fd = open (fil->fil_name, MODE_READ))
  		 == INVALID_HANDLE_VALUE)
+ #endif
  #endif
  		{
  #ifdef SUPERSERVER
Index: burp/mvol.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/mvol.c,v
retrieving revision 1.4
diff -c -r1.4 mvol.c
*** burp/mvol.c	2001/04/27 16:30:00	1.4
--- burp/mvol.c	2001/12/05 00:30:42
***************
*** 21,27 ****
--- 21,35 ----
   * Contributor(s): ______________________________________.
   */
  
+ #include "../jrd/common.h"
+ 
  #include "../jrd/ib_stdio.h"
+ 
+ /* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <stdlib.h>
  #include <string.h>
  #include <errno.h>
***************
*** 1018,1024 ****
--- 1026,1036 ----
      if((new_desc = MVOL_open(new_file,mode,OPEN_ALWAYS))
         == INVALID_HANDLE_VALUE)
  #else
+ #ifdef SOLARIS_64IO
+     if ((new_desc = open64 (new_file, mode, OPEN_MASK)) < 0)		
+ #else		
      if ((new_desc = open (new_file, mode, OPEN_MASK)) < 0)
+ #endif
  #endif /* WIN_NT */
  	{
  	BURP_print (222, new_file, 0, 0, 0, 0);
Index: burp/spit.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/spit.c,v
retrieving revision 1.2
diff -c -r1.2 spit.c
*** burp/spit.c	2000/08/30 04:21:35	1.2
--- burp/spit.c	2001/12/05 00:30:43
***************
*** 25,36 ****
   * Contributor(s): ______________________________________.
  */
  
  #include "../jrd/ib_stdio.h"
  #include <stdlib.h>
  #include <string.h>
  #include <errno.h>
  #include <assert.h>
- #include "../jrd/common.h"
  #include "../jrd/jrd.h"
  #include <stdarg.h>
  
--- 25,42 ----
   * Contributor(s): ______________________________________.
  */
  
+ #include "../jrd/common.h"
+ 
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <stdlib.h>
  #include <string.h>
  #include <errno.h>
  #include <assert.h>
  #include "../jrd/jrd.h"
  #include <stdarg.h>
  
***************
*** 667,674 ****
  
  		file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  		file_name	= fl_ptr->b_fil_name;
! 
  		output_fl_desc = open (file_name, MODE_WRITE, MASK);
  		if (output_fl_desc == -1)
  			{
  			free (io_buffer);
--- 673,683 ----
  
  		file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  		file_name	= fl_ptr->b_fil_name;
! #ifdef SOLARIS_64IO
! 		output_fl_desc = open64 (file_name, MODE_WRITE, MASK);
! #else
  		output_fl_desc = open (file_name, MODE_WRITE, MASK);
+ #endif
  		if (output_fl_desc == -1)
  			{
  			free (io_buffer);
***************
*** 746,753 ****
  
  						file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  						file_name	= fl_ptr->b_fil_name;
! 
  						output_fl_desc = open (file_name, MODE_WRITE, MASK);
  						if (output_fl_desc == -1)
  							{
  							free (io_buffer);
--- 755,765 ----
  
  						file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
  						file_name	= fl_ptr->b_fil_name;
! #ifdef SOLARIS_64IO
! 						output_fl_desc = open64 (file_name, MODE_WRITE, MASK);
! #else
  						output_fl_desc = open (file_name, MODE_WRITE, MASK);
+ #endif
  						if (output_fl_desc == -1)
  							{
  							free (io_buffer);
***************
*** 1073,1079 ****
--- 1085,1095 ----
  TEXT		*char_ptr1, *char_ptr2, num_arr[5], total_arr[5];
  HEADER_REC	hdr_rec;
  
+ #ifdef SOLARIS_64IO
+ input_fl_desc = open64 (file_name, MODE_READ);
+ #else
  input_fl_desc = open (file_name, MODE_READ);
+ #endif
  
  if (input_fl_desc == -1)
  	{
Index: jrd/common.h
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/common.h,v
retrieving revision 1.18
diff -c -r1.18 common.h
*** jrd/common.h	2001/10/31 04:50:39	1.18
--- jrd/common.h	2001/12/05 00:30:44
***************
*** 253,267 ****
   */
  #define NEED_IB_STDIO
  
  /* Define the following macro only on platforms which supply the library
     routine xdr_hyper.  For other platforms, omit the definition of this
     macro, and we will supply that function where we need it. */
  #define PLATFORM_SUPPLIES_XDR_HYPER
- 
- #ifdef SOLX86
- #define LSEEK_OFFSET_CAST (off_t)
- #endif
- 
  
  #define ATEXIT(c)       atexit (c)
  #define SETPGRP         setpgrp ()
--- 253,272 ----
   */
  #define NEED_IB_STDIO
  
+ /* Used for modules which can not be compiled with standard largefile
+    cflags (see prefix.solaris|solx86) due to problems with ib_stdio.
+    This is primarly used for utilities which do their own direct db file 
+    access and output to screen. 
+    Hopefully this can be cleaned up in FB2.
+ */   
+ #ifdef sparc
+ #define SOLARIS_64IO
+ #endif
+ 
  /* Define the following macro only on platforms which supply the library
     routine xdr_hyper.  For other platforms, omit the definition of this
     macro, and we will supply that function where we need it. */
  #define PLATFORM_SUPPLIES_XDR_HYPER
  
  #define ATEXIT(c)       atexit (c)
  #define SETPGRP         setpgrp ()
Index: jrd/unix.c
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/unix.c,v
retrieving revision 1.6
diff -c -r1.6 unix.c
*** jrd/unix.c	2001/10/29 19:45:30	1.6
--- jrd/unix.c	2001/12/05 00:30:45
***************
*** 126,132 ****
--- 126,136 ----
  extern int	errno;
  
  static void	close_marker_file (TEXT *);
+ #if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+ static FIL	seek_file (FIL, BDB, off_t *, STATUS *);
+ #else
  static FIL	seek_file (FIL, BDB, SLONG *, STATUS *);
+ #endif
  static FIL	setup_file (DBB, TEXT *, USHORT, int);
  static BOOLEAN	unix_error (TEXT *, FIL, STATUS, STATUS *);
  
***************
*** 326,331 ****
--- 330,339 ----
  flag = O_RDWR | O_CREAT | (overwrite ? O_TRUNC : O_EXCL) | O_BINARY;
  #endif
  
+ #ifdef SOLARIS_64IO
+ flag = flag | O_LARGEFILE;
+ #endif
+ 
  if ((desc = open (file_name, flag, MASK)) == -1)
      ERR_post (isc_io_error,
  	gds_arg_string, "open O_CREAT",
***************
*** 675,680 ****
--- 683,692 ----
  flag = O_RDWR | O_BINARY;
  #endif
  
+ #ifdef SOLARIS_64IO
+ flag = flag | O_LARGEFILE;
+ #endif
+ 
  for (i = 0; i < IO_RETRY; i++)
      {
  
***************
*** 732,739 ****
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size, offset;
  
  ISC_inhibit();
  
  if (file->fil_desc == -1)
--- 744,757 ----
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size; 
  
+ #if defined (SOLARIS_64IO) && defined (SUPERSERVER)
+ off_t offset;
+ #else
+ SLONG  offset;
+ #endif
+ 
  ISC_inhibit();
  
  if (file->fil_desc == -1)
***************
*** 833,840 ****
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size, offset;
  
  ISC_inhibit();
  
  if (file->fil_desc == -1)
--- 851,864 ----
   **************************************/
  DBB	dbb;
  SSHORT	i;
! SLONG	bytes, size;
  
+ #if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+ 	off_t  offset;
+ #else
+ 	SLONG offset;
+ #endif
+ 
  ISC_inhibit();
  
  if (file->fil_desc == -1)
***************
*** 945,951 ****
--- 969,979 ----
  static FIL seek_file (
      FIL		file,
      BDB		bdb,
+ #if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+     off_t	*offset,
+ #else
      SLONG	*offset,
+ #endif
      STATUS	*status_vector)
  {
  /**************************************
***************
*** 982,997 ****
  page -= file->fil_min_page - file->fil_fudge;
  
  #ifdef PREAD_PWRITE
! *offset = (SLONG) (page * dbb->dbb_page_size);
  #else
  lseek_offset = page;
  lseek_offset *= dbb->dbb_page_size;
  #ifndef UNIX_64_BIT_IO
  if (lseek_offset > MAX_SLONG)
  {
      return (FIL) unix_error ("lseek", file, isc_io_32bit_exceeded_err, status_vector);
  }
  #endif
  
  THD_MUTEX_LOCK (file->fil_mutex);
  
--- 1010,1027 ----
  page -= file->fil_min_page - file->fil_fudge;
  
  #ifdef PREAD_PWRITE
! *offset = (off_t) (page * dbb->dbb_page_size);
  #else
  lseek_offset = page;
  lseek_offset *= dbb->dbb_page_size;
  #ifndef UNIX_64_BIT_IO
+ #ifndef _LARGEFILE_SOURCE
  if (lseek_offset > MAX_SLONG)
  {
      return (FIL) unix_error ("lseek", file, isc_io_32bit_exceeded_err, status_vector);
  }
  #endif
+ #endif
  
  THD_MUTEX_LOCK (file->fil_mutex);
  
***************
*** 1146,1152 ****
      int fd,
      SCHAR *buf,
      SLONG nbytes,
!     SLONG offset)
  /**************************************
   *
   *	p r e a d
--- 1176,1187 ----
      int fd,
      SCHAR *buf,
      SLONG nbytes,
! #ifdef SOLARIS_64IO
! 	off_t
! #else
!     SLONG offset
! #endif
! )
  /**************************************
   *
   *	p r e a d
***************
*** 1183,1189 ****
      int fd,
      SCHAR *buf,
      SLONG nbytes,
!     SLONG offset)
  /**************************************
   *
   *	p w r i t e
--- 1218,1229 ----
      int fd,
      SCHAR *buf,
      SLONG nbytes,
! #ifdef SOLARIS_64IO
! 	off_t offset
! #else
!     SLONG offset
! #endif
! )
  /**************************************
   *
   *	p w r i t e
Index: utilities/dba.e
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/dba.e,v
retrieving revision 1.10
diff -c -r1.10 dba.e
*** utilities/dba.e	2001/10/29 19:45:30	1.10
--- utilities/dba.e	2001/12/05 00:30:47
***************
*** 27,32 ****
--- 27,37 ----
  #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <errno.h>
  #include <string.h>
  #include <fcntl.h>
***************
*** 2250,2256 ****
--- 2255,2265 ----
  fil->fil_fudge = 0;
  fil->fil_max_page = 0L;
  
+ #ifdef SOLARIS_64IO
+ if ((fil->fil_desc = open64 (file_name, O_RDONLY)) == -1)
+ #else
  if ((fil->fil_desc = open (file_name, O_RDONLY)) == -1)
+ #endif
      {
  #ifdef SUPERSERVER
      CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
***************
*** 2316,2322 ****
  
  page_number -= fil->fil_min_page - fil->fil_fudge;
  offset = ((UINT64)page_number) * ((UINT64)tddba->page_size);
! if (lseek (fil->fil_desc, offset, 0) == -1)
      {
  #ifdef SUPERSERVER
      CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
--- 2325,2335 ----
  
  page_number -= fil->fil_min_page - fil->fil_fudge;
  offset = ((UINT64)page_number) * ((UINT64)tddba->page_size);
! #ifdef SOLARIS_64IO
! if (lseek64 (fil->fil_desc, offset, SEEK_SET) == -1)
! #else
! if (lseek (fil->fil_desc, LSEEK_OFFSET_CAST offset, 0) == -1)
! #endif
      {
  #ifdef SUPERSERVER
      CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
Index: utilities/rebuild.c
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/rebuild.c,v
retrieving revision 1.2
diff -c -r1.2 rebuild.c
*** utilities/rebuild.c	2001/10/29 19:45:30	1.2
--- utilities/rebuild.c	2001/12/05 00:30:48
***************
*** 23,28 ****
--- 23,33 ----
  #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <errno.h>
  
  #include "../jrd/gds.h"
***************
*** 658,665 ****
   *	Open a database file.
   *
   **************************************/
! 
  if ((rbdb->rbdb_file.fil_file = open (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
      db_error (errno);
  }
  
--- 663,673 ----
   *	Open a database file.
   *
   **************************************/
! #ifdef SOLARIS_64IO
! if ((rbdb->rbdb_file.fil_file = open64 (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
! #else
  if ((rbdb->rbdb_file.fil_file = open (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
+ #endif
      db_error (errno);
  }
  
***************
*** 685,691 ****
--- 693,703 ----
  file = rbdb->rbdb_file.fil_file;
  
  offset = ((UINT64)page_number) * ((UINT64)rbdb->rbdb_page_size);
+ #ifdef SOLARIS_64IO
+ if (lseek64 (file, offset, SEEK_SET) == -1)
+ #else
  if (lseek (file, offset, 0) == -1)
+ #endif
      db_error (errno);
  
  for (p = (SCHAR*) rbdb->rbdb_buffer1, length = rbdb->rbdb_page_size; length > 0;)
***************
*** 726,732 ****
--- 738,748 ----
  fd = rbdb->rbdb_file.fil_file;
  
  offset = ((UINT64)page_number) * ((UINT64)page_size);
+ #ifdef SOLARIS_64IO
+ if (lseek64 (fd, offset, SEEK_SET) == -1)
+ #else
  if (lseek (fd, offset, 0) == -1)
+ #endif
      db_error (errno);
  if (write (fd, page, page_size) == -1)
      db_error (errno);
Index: utilities/sbc_print.c
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/sbc_print.c,v
retrieving revision 1.3
diff -c -r1.3 sbc_print.c
*** utilities/sbc_print.c	2001/10/29 19:45:30	1.3
--- utilities/sbc_print.c	2001/12/05 00:30:49
***************
*** 23,28 ****
--- 23,33 ----
  #include "../jrd/common.h"
  
  #include "../jrd/ib_stdio.h"
+ /* ib_stdio breaks with largefile compiling */
+ #ifdef SOLARIS_64IO
+ #define _LARGEFILE64_SOURCE
+ #endif
+ 
  #include <string.h>
  #include <fcntl.h>
  #include <errno.h>
***************
*** 858,864 ****
--- 863,873 ----
   *
   **************************************/
  
+ #idef SOLARIS_64IO
+ if ((file = open64 (file_name, 2)) == -1)
+ #else
  if ((file = open (file_name, 2)) == -1)
+ #endif
      db_error (errno);
  }
  
***************
*** 880,886 ****
--- 889,899 ----
  UINT64 offset;
  
  offset = ((UINT64)page_number) * ((UINT64)page_size);
+ #ifdef SOLARIS_64IO
+ if (lseek64 (file, offset, SEEK_SET) == -1)
+ #else
  if (lseek (file, offset, 0) == -1)
+ #endif
      db_error (errno);
  
  for (p = (SCHAR*) global_buffer, length = page_size; length > 0;)
Index: builds/original/prefix.solaris
===================================================================
RCS file: /cvsroot/firebird/interbase/builds/original/prefix.solaris,v
retrieving revision 1.2
diff -u -3 -r1.2 prefix.solaris
--- builds/original/prefix.solaris	2000/12/14 19:29:42	1.2
+++ builds/original/prefix.solaris	2001/12/05 00:38:50
@@ -14,6 +14,10 @@
 #
 # All Rights Reserved.
 # Contributor(s): ______________________________________.
+#
+# 31-Oct-2001  Neil McCalden  Add large file support
+#
+
 # Start of file prefix.solaris:	$(VERSION)	$(PLATFORM)
 .SUFFIXES: .c .e
 .e.c:
@@ -43,6 +47,9 @@
 DEV_SHRLIB_DIR=		-L source/jrd
 DEV_VERSION_FLAG=	-DDEV_BUILD
 DEV_CFLAGS=		-g -xstrconst
+
+# Also set SOLARIS_64IO in common.h
+LARGEFILE_CFLAGS=`/usr/bin/getconf LFS_CFLAGS`
 
 DEBUG_OBJECTS=		$($(VERSION)_DEBUG_OBJECTS)
 SHRLIB_DIR=		$($(VERSION)_SHRLIB_DIR)
Index: burp/burp.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/burp.c,v
retrieving revision 1.7
diff -u -3 -r1.7 burp.c
--- burp/burp.c	2001/08/09 00:27:39	1.7
+++ burp/burp.c	2001/12/05 00:38:52
@@ -25,12 +25,18 @@
  */
 
 
+#include "../jrd/common.h"
 
 #include "../jrd/ib_stdio.h"
+/* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+#ifdef SOLARIS_64IO
+#define _LARGEFILE64_SOURCE
+#endif
+
+
 #include <stdlib.h>
 #include <string.h>
 #include <signal.h>
-#include "../jrd/common.h"
 #include "../jrd/ibase.h"
 #include <stdarg.h>
 #include "../jrd/ibsetjmp.h"
@@ -1765,7 +1771,11 @@
 	else
 #ifndef mpexl
 #ifndef WIN_NT
+#ifdef SOLARIS_64IO
+	    if ((fil->fil_fd = open64 (fil->fil_name, MODE_WRITE, MASK)) == -1)
+#else
 	    if ((fil->fil_fd = open (fil->fil_name, MODE_WRITE, MASK)) == -1)
+#endif  
 #else
 	    if ((fil->fil_fd = MVOL_open (fil->fil_name, MODE_WRITE,
 		 CREATE_ALWAYS)) == INVALID_HANDLE_VALUE)
@@ -1914,8 +1924,12 @@
     if ((fil->fil_fd =  MVOL_open (fil->fil_name, MODE_READ, OPEN_EXISTING))
 	 == INVALID_HANDLE_VALUE)
 #else
+#ifdef SOLARIS_64IO
+    if ((fil->fil_fd = open64 (fil->fil_name, MODE_READ)) == INVALID_HANDLE_VALUE)		
+#else
     if ((fil->fil_fd = open (fil->fil_name, MODE_READ)) == INVALID_HANDLE_VALUE)
 #endif
+#endif
 	{
 	BURP_error (65, fil->fil_name, 0, 0, 0, 0);  /* msg 65 can't open backup file %s */
 	return QUIT;
@@ -1954,8 +1968,13 @@
 	    if ((fil->fil_fd =  MVOL_open (fil->fil_name, MODE_READ,
 		 OPEN_EXISTING)) == INVALID_HANDLE_VALUE)
 #else
+#ifdef SOLARIS_64IO
+	    if ((fil->fil_fd = open64 (fil->fil_name, MODE_READ))
+		 == INVALID_HANDLE_VALUE)		
+#else			
 	    if ((fil->fil_fd = open (fil->fil_name, MODE_READ))
 		 == INVALID_HANDLE_VALUE)
+#endif
 #endif
 		{
 #ifdef SUPERSERVER
Index: burp/mvol.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/mvol.c,v
retrieving revision 1.4
diff -u -3 -r1.4 mvol.c
--- burp/mvol.c	2001/04/27 16:30:00	1.4
+++ burp/mvol.c	2001/12/05 00:38:54
@@ -21,7 +21,15 @@
  * Contributor(s): ______________________________________.
  */
 
+#include "../jrd/common.h"
+
 #include "../jrd/ib_stdio.h"
+
+/* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+#ifdef SOLARIS_64IO
+#define _LARGEFILE64_SOURCE
+#endif
+
 #include <stdlib.h>
 #include <string.h>
 #include <errno.h>
@@ -1018,7 +1026,11 @@
     if((new_desc = MVOL_open(new_file,mode,OPEN_ALWAYS))
        == INVALID_HANDLE_VALUE)
 #else
+#ifdef SOLARIS_64IO
+    if ((new_desc = open64 (new_file, mode, OPEN_MASK)) < 0)		
+#else		
     if ((new_desc = open (new_file, mode, OPEN_MASK)) < 0)
+#endif
 #endif /* WIN_NT */
 	{
 	BURP_print (222, new_file, 0, 0, 0, 0);
Index: burp/spit.c
===================================================================
RCS file: /cvsroot/firebird/interbase/burp/spit.c,v
retrieving revision 1.2
diff -u -3 -r1.2 spit.c
--- burp/spit.c	2000/08/30 04:21:35	1.2
+++ burp/spit.c	2001/12/05 00:38:54
@@ -25,12 +25,18 @@
  * Contributor(s): ______________________________________.
 */
 
+#include "../jrd/common.h"
+
 #include "../jrd/ib_stdio.h"
+/* ib_stdio breaks with largefile compiling - use transitional mixed 32/64 */
+#ifdef SOLARIS_64IO
+#define _LARGEFILE64_SOURCE
+#endif
+
 #include <stdlib.h>
 #include <string.h>
 #include <errno.h>
 #include <assert.h>
-#include "../jrd/common.h"
 #include "../jrd/jrd.h"
 #include <stdarg.h>
 
@@ -667,8 +673,11 @@
 
 		file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
 		file_name	= fl_ptr->b_fil_name;
-
+#ifdef SOLARIS_64IO
+		output_fl_desc = open64 (file_name, MODE_WRITE, MASK);
+#else
 		output_fl_desc = open (file_name, MODE_WRITE, MASK);
+#endif
 		if (output_fl_desc == -1)
 			{
 			free (io_buffer);
@@ -746,8 +755,11 @@
 
 						file_size	= fl_ptr->b_fil_size - HEADER_REC_LEN;
 						file_name	= fl_ptr->b_fil_name;
-
+#ifdef SOLARIS_64IO
+						output_fl_desc = open64 (file_name, MODE_WRITE, MASK);
+#else
 						output_fl_desc = open (file_name, MODE_WRITE, MASK);
+#endif
 						if (output_fl_desc == -1)
 							{
 							free (io_buffer);
@@ -1073,7 +1085,11 @@
 TEXT		*char_ptr1, *char_ptr2, num_arr[5], total_arr[5];
 HEADER_REC	hdr_rec;
 
+#ifdef SOLARIS_64IO
+input_fl_desc = open64 (file_name, MODE_READ);
+#else
 input_fl_desc = open (file_name, MODE_READ);
+#endif
 
 if (input_fl_desc == -1)
 	{
Index: jrd/common.h
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/common.h,v
retrieving revision 1.18
diff -u -3 -r1.18 common.h
--- jrd/common.h	2001/10/31 04:50:39	1.18
+++ jrd/common.h	2001/12/05 00:38:56
@@ -253,15 +253,20 @@
  */
 #define NEED_IB_STDIO
 
+/* Used for modules which can not be compiled with standard largefile
+   cflags (see prefix.solaris|solx86) due to problems with ib_stdio.
+   This is primarly used for utilities which do their own direct db file 
+   access and output to screen. 
+   Hopefully this can be cleaned up in FB2.
+*/   
+#ifdef sparc
+#define SOLARIS_64IO
+#endif
+
 /* Define the following macro only on platforms which supply the library
    routine xdr_hyper.  For other platforms, omit the definition of this
    macro, and we will supply that function where we need it. */
 #define PLATFORM_SUPPLIES_XDR_HYPER
-
-#ifdef SOLX86
-#define LSEEK_OFFSET_CAST (off_t)
-#endif
-
 
 #define ATEXIT(c)       atexit (c)
 #define SETPGRP         setpgrp ()
Index: jrd/unix.c
===================================================================
RCS file: /cvsroot/firebird/interbase/jrd/unix.c,v
retrieving revision 1.6
diff -u -3 -r1.6 unix.c
--- jrd/unix.c	2001/10/29 19:45:30	1.6
+++ jrd/unix.c	2001/12/05 00:38:57
@@ -126,7 +126,11 @@
 extern int	errno;
 
 static void	close_marker_file (TEXT *);
+#if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+static FIL	seek_file (FIL, BDB, off_t *, STATUS *);
+#else
 static FIL	seek_file (FIL, BDB, SLONG *, STATUS *);
+#endif
 static FIL	setup_file (DBB, TEXT *, USHORT, int);
 static BOOLEAN	unix_error (TEXT *, FIL, STATUS, STATUS *);
 
@@ -326,6 +330,10 @@
 flag = O_RDWR | O_CREAT | (overwrite ? O_TRUNC : O_EXCL) | O_BINARY;
 #endif
 
+#ifdef SOLARIS_64IO
+flag = flag | O_LARGEFILE;
+#endif
+
 if ((desc = open (file_name, flag, MASK)) == -1)
     ERR_post (isc_io_error,
 	gds_arg_string, "open O_CREAT",
@@ -675,6 +683,10 @@
 flag = O_RDWR | O_BINARY;
 #endif
 
+#ifdef SOLARIS_64IO
+flag = flag | O_LARGEFILE;
+#endif
+
 for (i = 0; i < IO_RETRY; i++)
     {
 
@@ -732,8 +744,14 @@
  **************************************/
 DBB	dbb;
 SSHORT	i;
-SLONG	bytes, size, offset;
+SLONG	bytes, size; 
 
+#if defined (SOLARIS_64IO) && defined (SUPERSERVER)
+off_t offset;
+#else
+SLONG  offset;
+#endif
+
 ISC_inhibit();
 
 if (file->fil_desc == -1)
@@ -833,8 +851,14 @@
  **************************************/
 DBB	dbb;
 SSHORT	i;
-SLONG	bytes, size, offset;
+SLONG	bytes, size;
 
+#if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+	off_t  offset;
+#else
+	SLONG offset;
+#endif
+
 ISC_inhibit();
 
 if (file->fil_desc == -1)
@@ -945,7 +969,11 @@
 static FIL seek_file (
     FIL		file,
     BDB		bdb,
+#if defined(SOLARIS_64IO) && defined(SUPERSERVER)
+    off_t	*offset,
+#else
     SLONG	*offset,
+#endif
     STATUS	*status_vector)
 {
 /**************************************
@@ -982,16 +1010,18 @@
 page -= file->fil_min_page - file->fil_fudge;
 
 #ifdef PREAD_PWRITE
-*offset = (SLONG) (page * dbb->dbb_page_size);
+*offset = (off_t) (page * dbb->dbb_page_size);
 #else
 lseek_offset = page;
 lseek_offset *= dbb->dbb_page_size;
 #ifndef UNIX_64_BIT_IO
+#ifndef _LARGEFILE_SOURCE
 if (lseek_offset > MAX_SLONG)
 {
     return (FIL) unix_error ("lseek", file, isc_io_32bit_exceeded_err, status_vector);
 }
 #endif
+#endif
 
 THD_MUTEX_LOCK (file->fil_mutex);
 
@@ -1146,7 +1176,12 @@
     int fd,
     SCHAR *buf,
     SLONG nbytes,
-    SLONG offset)
+#ifdef SOLARIS_64IO
+	off_t
+#else
+    SLONG offset
+#endif
+)
 /**************************************
  *
  *	p r e a d
@@ -1183,7 +1218,12 @@
     int fd,
     SCHAR *buf,
     SLONG nbytes,
-    SLONG offset)
+#ifdef SOLARIS_64IO
+	off_t offset
+#else
+    SLONG offset
+#endif
+)
 /**************************************
  *
  *	p w r i t e
Index: utilities/dba.e
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/dba.e,v
retrieving revision 1.10
diff -u -3 -r1.10 dba.e
--- utilities/dba.e	2001/10/29 19:45:30	1.10
+++ utilities/dba.e	2001/12/05 00:38:58
@@ -27,6 +27,11 @@
 #include "../jrd/common.h"
 
 #include "../jrd/ib_stdio.h"
+/* ib_stdio breaks with largefile compiling */
+#ifdef SOLARIS_64IO
+#define _LARGEFILE64_SOURCE
+#endif
+
 #include <errno.h>
 #include <string.h>
 #include <fcntl.h>
@@ -2250,7 +2255,11 @@
 fil->fil_fudge = 0;
 fil->fil_max_page = 0L;
 
+#ifdef SOLARIS_64IO
+if ((fil->fil_desc = open64 (file_name, O_RDONLY)) == -1)
+#else
 if ((fil->fil_desc = open (file_name, O_RDONLY)) == -1)
+#endif
     {
 #ifdef SUPERSERVER
     CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
@@ -2316,7 +2325,11 @@
 
 page_number -= fil->fil_min_page - fil->fil_fudge;
 offset = ((UINT64)page_number) * ((UINT64)tddba->page_size);
-if (lseek (fil->fil_desc, offset, 0) == -1)
+#ifdef SOLARIS_64IO
+if (lseek64 (fil->fil_desc, offset, SEEK_SET) == -1)
+#else
+if (lseek (fil->fil_desc, LSEEK_OFFSET_CAST offset, 0) == -1)
+#endif
     {
 #ifdef SUPERSERVER
     CMD_UTIL_put_svc_status (tddba->dba_service_blk->svc_status,
Index: utilities/rebuild.c
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/rebuild.c,v
retrieving revision 1.2
diff -u -3 -r1.2 rebuild.c
--- utilities/rebuild.c	2001/10/29 19:45:30	1.2
+++ utilities/rebuild.c	2001/12/05 00:39:00
@@ -23,6 +23,11 @@
 #include "../jrd/common.h"
 
 #include "../jrd/ib_stdio.h"
+/* ib_stdio breaks with largefile compiling */
+#ifdef SOLARIS_64IO
+#define _LARGEFILE64_SOURCE
+#endif
+
 #include <errno.h>
 
 #include "../jrd/gds.h"
@@ -658,8 +663,11 @@
  *	Open a database file.
  *
  **************************************/
-
+#ifdef SOLARIS_64IO
+if ((rbdb->rbdb_file.fil_file = open64 (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
+#else
 if ((rbdb->rbdb_file.fil_file = open (rbdb->rbdb_file.fil_name, O_RDWR, 0)) == -1)
+#endif
     db_error (errno);
 }
 
@@ -685,7 +693,11 @@
 file = rbdb->rbdb_file.fil_file;
 
 offset = ((UINT64)page_number) * ((UINT64)rbdb->rbdb_page_size);
+#ifdef SOLARIS_64IO
+if (lseek64 (file, offset, SEEK_SET) == -1)
+#else
 if (lseek (file, offset, 0) == -1)
+#endif
     db_error (errno);
 
 for (p = (SCHAR*) rbdb->rbdb_buffer1, length = rbdb->rbdb_page_size; length > 0;)
@@ -726,7 +738,11 @@
 fd = rbdb->rbdb_file.fil_file;
 
 offset = ((UINT64)page_number) * ((UINT64)page_size);
+#ifdef SOLARIS_64IO
+if (lseek64 (fd, offset, SEEK_SET) == -1)
+#else
 if (lseek (fd, offset, 0) == -1)
+#endif
     db_error (errno);
 if (write (fd, page, page_size) == -1)
     db_error (errno);
Index: utilities/sbc_print.c
===================================================================
RCS file: /cvsroot/firebird/interbase/utilities/sbc_print.c,v
retrieving revision 1.3
diff -u -3 -r1.3 sbc_print.c
--- utilities/sbc_print.c	2001/10/29 19:45:30	1.3
+++ utilities/sbc_print.c	2001/12/05 00:39:00
@@ -23,6 +23,11 @@
 #include "../jrd/common.h"
 
 #include "../jrd/ib_stdio.h"
+/* ib_stdio breaks with largefile compiling */
+#ifdef SOLARIS_64IO
+#define _LARGEFILE64_SOURCE
+#endif
+
 #include <string.h>
 #include <fcntl.h>
 #include <errno.h>
@@ -858,7 +863,11 @@
  *
  **************************************/
 
+#idef SOLARIS_64IO
+if ((file = open64 (file_name, 2)) == -1)
+#else
 if ((file = open (file_name, 2)) == -1)
+#endif
     db_error (errno);
 }
 
@@ -880,7 +889,11 @@
 UINT64 offset;
 
 offset = ((UINT64)page_number) * ((UINT64)page_size);
+#ifdef SOLARIS_64IO
+if (lseek64 (file, offset, SEEK_SET) == -1)
+#else
 if (lseek (file, offset, 0) == -1)
+#endif
     db_error (errno);
 
 for (p = (SCHAR*) global_buffer, length = page_size; length > 0;)
